libbeat/autodiscover/providers/kubernetes/kubernetes.go

```diff
@@ -154,3 +181,152 @@ func (p *Provider) publish(event bus.Event) {
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return em, nil
+}
+
+func NewLeaderElectionManager(
+	uuid uuid.UUID,
+	cfg *Config,
+	client k8s.Interface,
+	startLeading func(uuid string, eventID string),
+	stopLeading func(uuid string, eventID string),
+	logger *logp.Logger,
+) (EventManager, error) {
+	lem := &leaderElectionManager{logger: logger}
+	var id string
+	if cfg.Node != "" {
+		id = "beats-leader-" + cfg.Node
+	} else {
+		id = "beats-leader-" + uuid.String()
+	}
+	lease := metav1.ObjectMeta{
+		Name:      cfg.LeaderLease,
+		Namespace: "default",
+	}
+	metaUID := lease.GetObjectMeta().GetUID()
+	lem.leaderElection = leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: lease,
+			Client:    client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				logger.Debugf("leader election lock GAINED, id %v", id)
+				eventID := fmt.Sprintf("%v-%v", metaUID, time.Now().UnixNano())
+				startLeading(uuid.String(), eventID)
+			},
+			OnStoppedLeading: func() {
+				logger.Debugf("leader election lock LOST, id %v", id)
+				eventID := fmt.Sprintf("%v-%v", metaUID, time.Now().UnixNano())
+				stopLeading(uuid.String(), eventID)
+			},
+		},
+	}
+	return lem, nil
+}
+
+// Start for EventManager interface.
+func (p *eventerManager) Start() {
+	if err := p.eventer.Start(); err != nil {
+		p.logger.Errorf("Error starting kubernetes autodiscover provider: %s", err)
+	}
+}
+
+// Stop signals the stop channel to force the watch loop routine to stop.
+func (p *eventerManager) Stop() {
+	p.eventer.Stop()
+}
+
+// GenerateHints for EventManager interface.
+func (p *eventerManager) GenerateHints(event bus.Event) bus.Event {
+	return p.eventer.GenerateHints(event)
+}
+
+// Start for EventManager interface.
+func (p *leaderElectionManager) Start() {
+	ctx, cancel := context.WithCancel(context.Background())
+	p.cancelLeaderElection = cancel
+	p.startLeaderElector(ctx, p.leaderElection)
+}
+
+// Stop signals the stop channel to force the leader election loop routine to stop.
+func (p *leaderElectionManager) Stop() {
+	if p.cancelLeaderElection != nil {
+		p.cancelLeaderElection()
+	}
+}
+
+// GenerateHints for EventManager interface.
+func (p *leaderElectionManager) GenerateHints(event bus.Event) bus.Event {
+	return event
+}
+
+// startLeaderElector starts a Leader Elector in the background with the provided config
+func (p *leaderElectionManager) startLeaderElector(ctx context.Context, lec leaderelection.LeaderElectionConfig) {
+	le, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		p.logger.Errorf("leader election lock GAINED, id %v", err)
+	}
+	if lec.WatchDog != nil {
```

Comment:
What is this watchdog? we are never setting it.

Reply:
This was mostly because I tried to "copy" the `RunOrDie` of https://github.com/kubernetes/client-go/blob/36233866f1c7c0ad3bdac1fc466cb5de3746cfa2/tools/leaderelection/leaderelection.go#L219, because of an async problem described at https://github.com/kubernetes/client-go/issues/837 .

Good catch I will remove it, it's redundant.